integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/BaseVisitorFactory.java

```diff
@@ -0,0 +1,74 @@
+package io.openlineage.spark.agent.lifecycle;
+
+import io.openlineage.client.OpenLineage;
+import io.openlineage.spark.agent.lifecycle.plan.AppendDataVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.BigQueryNodeVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.CommandPlanVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.InsertIntoDataSourceDirVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.InsertIntoDataSourceVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.InsertIntoDirVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.InsertIntoHadoopFsRelationVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.InsertIntoHiveDirVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.InsertIntoHiveTableVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.LogicalRDDVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.LogicalRelationVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.QueryPlanVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.SaveIntoDataSourceCommandVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.VisitorFactory;
+import io.openlineage.spark.agent.lifecycle.plan.wrapper.InputDatasetVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.wrapper.OutputDatasetVisitor;
+import io.openlineage.spark.agent.lifecycle.plan.wrapper.OutputDatasetWithMetadataVisitor;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.spark.sql.SQLContext;
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
+
+public abstract class BaseVisitorFactory implements VisitorFactory {
```

Comment:
We should reduce visibility on all of these internal classes and interfaces. There's no need to expose these internals outside of the library. Let's move `VisitorFactory` into the same package as this class and make both package scoped. The Spark-version specific subclasses must also be moved into this same package and named to avoid conflicts (e.g., `Spark3VisitorFactoryImpl`).

Reply:
as I remember we agree that spark2 and spark3 will have different package name, like
```
io.openlineage.spark2.agent.lifecycle.plan
io.openlineage.spark3.agent.lifecycle.plan
```
Are you saying that we should have package `io.openlineage.spark.agent.lifecycle.plan` for both spark2 and spark3
If we have different packages I can't use package-private scope

Reply:
We can use the `io.openlineage.spark.agent.lifecycle.plan` package for the `VisitorFactory` and its implementations. The Spark2 and Spark3 versions can use different class names, rather than relying on the package name. Otherwise, we need to make all these classes public and I'd prefer to keep them hidden so people don't build on them for whatever reason.